### PR TITLE
fix: [CDS-77211]: adds environment.infrastructureDefinitions to KEYS_TO_IGNORE set

### PR DESCRIPTION
**Pull request by sandeshsalunkhegh**
fix: [CDS-77211]: adds environment.infrastructureDefinitions to KEYS_TO_IGNORE set and adds unit tests to test runtime inputs validation for this scenario

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

[CDS-77211]: https://harness.atlassian.net/browse/CDS-77211?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51979)

### CODE REVIEW

**Comment by sandeshsalunkhegh** [comment]
trigger messagecheck

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51979)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`template-service/service/src/main/java/io/harness/template/helpers/InputsValidator.java` line 60 at r2](https://reviewable.io/reviews/harness/harness-core/51979#-Nc_tHkS0hWt6VyU_JvC:-Nc_tHkS0hWt6VyU_JvD:b4ns1nc) ([raw file](https://github.com/harness/harness-core/blob/583557c7a6a4eb5e349b5faf53781761b705e850/template-service/service/src/main/java/io/harness/template/helpers/InputsValidator.java#L60)):*
> ```Java
>   //  them is implemented.
>   private final Set<String> KEYS_TO_IGNORE = Set.of("service.serviceInputs", "environment.environmentInputs",
>       "environment.serviceOverrideInputs", "codebase.repoName", "environment.infrastructureDefinitions");
> ```

@tchaurasiya  could you please take a look at this ?  
If we ignore `environment.infrastructureDefinitions`, will reconcille happen for infra inputs ?

**Comment by tchaurasiya** [comment]
> 

https://harness.slack.com/archives/CJFLDARK9/p1692853664823169?thread_ts=1692807361.532819&cid=CJFLDARK9
Discussed here

File: 953-yaml-commons/src/main/java/io/harness/pms/merger/helpers/RuntimeInputsValidator.java [hunk 1]
```diff
@@ -159,15 +159,15 @@ public boolean validateInputsAgainstSourceNode(String nodeToValidateYaml, String
...
-
-          // This is to handle if sourceYaml has environment.infrastructureDefinitions as input but user choose
-          // useFromStage in nodeToRefresh
-          Optional<FQN> nodeFQNOneOfForInfraDefinition =
-              isNodeToValidationKeyIsOneOfForInfraDefinition(nodeToValidateFqnToValueMap, key);
-          if (nodeFQNOneOfForInfraDefinition.isPresent()) {
-            nodeToValidateFqnToValueMap.remove(nodeFQNOneOfForInfraDefinition.get());
-            continue;
-          }
+          //           TODO: Please uncomment below code-block to handle environment.infrastructureDefinitions once the
```
Comment (tchaurasiya) [inline 953-yaml-commons/src/main/java/io/harness/pms/merger/helpers/RuntimeInputsValidator.java:162]: 1. Lets add junit for the affected scenarios.

2. Please test below scenario
sourceYaml (templateYaml has infrastructureDefinitions : <+input> ) and nodeToValidateYaml (pipelineYaml does not have infrastructureDefinitions : <+input> somehow, (user can reach to this state, by removing manually from yaml or deleting infra from pipeline studio page)
as mentioned https://harness.slack.com/archives/CJFLDARK9/p1692853664823169?thread_ts=1692807361.532819&cid=CJFLDARK9

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

File: 953-yaml-commons/src/main/java/io/harness/pms/merger/helpers/RuntimeInputsValidator.java (hunk 1 above)
Comment (sandeshsalunkhegh) [reply 953-yaml-commons/src/main/java/io/harness/pms/merger/helpers/RuntimeInputsValidator.java]: > 1. Lets add junit for the affected scenarios.
> 2. Please test below scenario
>    sourceYaml (templateYaml has infrastructureDefinitions : <+input> ) and nodeToValidateYaml (pipelineYaml does not have infrastructureDefinitions : <+input> somehow, (user can reach to this state, by removing manually from yaml or deleting infra from pipeline studio page)
>    as mentioned https://harness.slack.com/archives/CJFLDARK9/p1692853664823169?thread_ts=1692807361.532819&cid=CJFLDARK9

Added unit tests for the same, and other different possible scenarios as well.

File: 953-yaml-commons/src/test/java/io/harness/pms/merger/helpers/RuntimeInputsValidatorTest.java [hunk 2]
```diff
@@ -264,13 +265,13 @@ public void testValidateInputsWithUseFromStage() throws IOException {
   @Category(UnitTests.class)
   public void testValidateWithUseFromStageMultiService() throws IOException {
     Set<String> KEYS_TO_IGNORE = Set.of("service.serviceInputs", "environment.environmentInputs",
-        "environment.serviceOverrideInputs", "codebase.repoName");
```
Comment (tchaurasiya) [inline 953-yaml-commons/src/test/java/io/harness/pms/merger/helpers/RuntimeInputsValidatorTest.java:267]: Why are we modifying test case? Existing test case should work with new changes.

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

File: 953-yaml-commons/src/test/java/io/harness/pms/merger/helpers/RuntimeInputsValidatorTest.java (hunk 2 above)
Comment (sandeshsalunkhegh) [reply 953-yaml-commons/src/test/java/io/harness/pms/merger/helpers/RuntimeInputsValidatorTest.java]: @tchaurasiya I have reverted this test.

**Pull request closed without merging**